apps/editing-toolkit/editing-toolkit-plugin/tutorials/class-wpcom-tutorials.php

```diff
@@ -0,0 +1,253 @@
+			// Default status is pending.
+			$task['status'] = 'pending';
+			if ( isset( $statuses[ $task['id'] ] ) ) {
+				$task['status'] = $statuses[ $task['id'] ];
+			}
+			$tasks_with_statuses[] = $task;
+		}
+
+		return $tasks_with_statuses;
+	}
+
+	/**
+	 * Gets the completion statuses for a specific Tutorial.
+	 *
+	 * @param string $id Tutorial ID.
+	 * @return array     Completion statuses for Tutorial ID.
+	 */
+	public function get_statuses( $id ) {
+		$statuses = $this->load_persisted_statuses();
+		return isset( $statuses[ $id ] ) ? $statuses[ $id ] : array();
+	}
+
+	/**
+	 * Loads all Tutorial statuses from the database.
+	 *
+	 * @return array All Tutorial completion statuses.
+	 */
+	private function load_persisted_statuses() {
+		if ( function_exists( 'get_user_attribute' ) ) {
+			return get_user_attribute( get_current_user_id(), 'wpcom_tutorials_status' );
+		}
+		return (array) get_user_meta( get_current_user_id(), 'wpcom_tutorials_status', true );
+	}
+
+	/**
+	 * Marks a task with a status.
+	 *
+	 * @param string $tutorial_id Tutorial ID.
+	 * @param string $task_id     Task ID.
+	 * @param string $status      Status for the task. Allowed: complete|skipped|pending.
+	 * @return bool               True if status is updated, false otherwise.
+	 */
+	public function mark_task( $tutorial_id, $task_id, $status ) {
+		if ( ! $this->validate_status( $status ) ) {
+			return false;
+		}
+		if ( ! $this->task_exists( $tutorial_id, $task_id ) ) {
+			return false;
+		}
+
+		$statuses = $this->load_persisted_statuses();
+		if ( ! isset( $statuses[ $tutorial_id ] ) ) {
+			$statuses[ $tutorial_id ] = array();
+		}
+		// pending is the natural state with no stored status.
+		if ( 'pending' === $status ) {
+			if ( isset( $statuses[ $tutorial_id ][ $task_id ] ) ) {
+				unset( $statuses[ $tutorial_id ][ $task_id ] );
+			}
+		} else {
+			$statuses[ $tutorial_id ][ $task_id ] = $status;
+		}
+
+		return $this->persist_statuses( $statuses );
+	}
+
+	/**
+	 * Checks if a task exists for a registered tutorial.
+	 *
+	 * @param string $tutorial_id A tutorial ID.
+	 * @param string $task_id     A task ID.
+	 * @return bool True if it exists, otherwise false.
+	 */
+	private function task_exists( $tutorial_id, $task_id ) {
+		$tutorial = $this->get_tutorial( $tutorial_id );
+		foreach ( $tutorial['tasks'] as $task ) {
+			if ( $task_id === $task['id'] ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Persist statuses to the DB, using the platform-appropriate tech
+	 *
+	 * @param array $statuses All statuses to persist.
+	 * @return bool           True if persisting succeeded, otherwise false.
+	 */
+	public function persist_statuses( $statuses ) {
+		if ( function_exists( 'update_user_attribute' ) ) {
+			return update_user_attribute( get_current_user_id(), 'wpcom_tutorials_status', $statuses );
+		}
+		return update_user_meta( get_current_user_id(), 'wpcom_tutorials_status', $statuses );
```

Comment:
Im thinking about more questions regarding user attributes and meta.

* When a site transfers to atomic, are the attributes automatically migrated into meta? Or do we need to create a follow up to handle that for this specific case?
* I assume the meta is specific to the atomic site. So if a user had 5 different atomic sites, the tutorial statuses across those sites wouldn't be in sync? But similarly if a user had 5 simple sites, the tutorial statuses would be in sync? Is that accurate?